lightly/models/modules/smog.py

```diff
@@ -0,0 +1,85 @@
+from typing import Optional
+
+import torch
+import torch.nn as nn
+
+
+class SMoG(nn.Module):
+    """SMoG module for synchronous momentum grouping.
+
+    Attributes:
+        n_groups:
+            Number of groups to cluster.
+        dim:
+            Dimensionality of the group representations.
+        beta:
+            Momentum update factor of the group representations.
+        device:
+            Device to select.
+    
+    """
+
+    def __init__(
+        self, n_groups: int, dim: int, beta: float, device: Optional[str] = None
+    ):
+        super(SMoG, self).__init__()
+        self.n_groups = n_groups
+        self.dim = dim
+        self.group_features = torch.rand(n_groups, dim).to(device)
```

Comment:
I would maybe make `self.group_features` a parameter of the module. Then you can also easily save it / move it to other devices.

Reply:
This doesn't work because it requires a gradient but only for the "added" part.

Reply:
I fear I don't understand. Parameters can have gradients. And you can always disable them temporarily with `torch.no_grad`.

Reply:
I'm trying out some stuff. The problem is that you do a momentum update and you only want the gradient only for the "update" part.